### ISSUE TITLE
ktlint 1.1.0 doesn't respect disabled discouraged-comment-location rule in .editorconfig

### ISSUE THREAD

**Issue by Spikhalskiy** [pinterest/ktlint#2445]
We don't want to enforce comments location in our project because we like the following code style:

```
        val object = Object(
            someParam1 = 10000, // comment here about why
            someParam2 = 100, // comment here about why
            sameParam3 = emptyList(), // comment here about why
        )
```
It's compact and nice to work with, but discouraged by ktlint.

To disable the `A comment in a 'value_argument_list' is only allowed when placed on a separate line (cannot be auto-corrected)` warnings for such usages, our `.editorconfig` file looks the following way:
```
root = true

[*]
ktlint_code_style = intellij_idea
ktlint_standard_discouraged-comment-location = disabled
```

It works for ktlint 1.0.1, but ignored when we try to update ktlint to 1.1.0.

### Additional context

We are using org.jlleitschuh.gradle.ktlint:12.0.3 gradle plugin configured the following way:
```
    ktlint {
        version.set(libs.versions.ktlint)
        coloredOutput.set(true)
        android.set(false)
        ignoreFailures.set(false)
        filter {
            exclude("**/generated/**")
            exclude { element -> element.file.path.contains("generated/") }
        }
    }
```

**Comment by eygraber** [comment]
Looks like a new rule was added for this - `value-argument-comment`

Unfortunately disabling it results in:

> Rule with id 'RuleId(value=standard:argument-list-wrapping)' requires rule with id 'RuleId(value=standard:value-argument-comment)' to be loaded

At this rate I'll have most ktlint rules disabled by 2025 :disappointed:

**Comment by Spikhalskiy** [comment]
It would be nice to see the mention of the new `value-argument-comment` and `value-parameter-comment` rules
1. in docs: https://pinterest.github.io/ktlint/1.1.0/rules/standard/
2. in release notes: https://github.com/pinterest/ktlint/releases/tag/1.1.0, ideally including some kind of migration guide?

Also, yeah, it looks like disabling  `value-argument-comment` requires disabling `argument-list-wrapping`, and disabling  `value-parameter-comment` requires disabling of `function-signature` rule.
It's not fully clear why. 
For the `argument-list-wrapping`, why can't we enforce that every parameter is on a separate line if there are trailing comments?
For the `function-signature`, why can't we still place into one line the function arguments with short comments? But I see that this may require changing line comments to block comments.

**Comment by paul-dingemans** [comment]
Sorry for not clearly communicating that the `discouraged-comment-location` has been splitted into separate rules:
* `type-argument-comment`
* `type-parameter-comment`
* `value-argument-comment`
* `value-parameter-comment`

Reason for this is to provide better insights why certain rules are blocked.

With ktlint `1.0.1`, disabling `discouraged-comment-location` was only possible when also disabling:
* `chain-method-continuation`
* `class-signature`
* `if-else-wrapping`
From a user perspective it makes what the relation is between those rules.

With ktlint `1.1.0`, disabling `value-argument-comment` is only possible when also disabling:
* `argument-list-wrapping`
* `chain-method-continuation`
The `chain-method-continuation` should not have been blocked. From a functional perspective, and from the unit test perspective, there seems no reason to block it.

Intuitively the relation between `argument-list-wrapping` and `value-argument-comment` is not too bad even in case you might not agree with disallowing comments in and after value arguments. It is a common pattern to use. But syntaxtically it is not correct as the comment is place *after* the semi colon while it refers to the element *before* the comment. Not only ktlint has troubles to process that correctly but also IntelliJ IDEA does not deal with it correctly as can be seen in screenshots below:

We start with this code:
![Screenshot 2023-12-19 at 21 25 44](https://github.com/pinterest/ktlint/assets/5195292/f393f0bf-52fe-40f7-a203-386e158bf279)


Open 'Change signature' and select 'sameParam3':
![Screenshot 2023-12-19 at 21 26 14](https://github.com/pinterest/ktlint/assets/5195292/940cb554-453f-4bb2-9e66-8f2909c9f16e)


Now move 'sameParam3' one position up and press 'Refactor':
![Screenshot 2023-12-19 at 21 26 24](https://github.com/pinterest/ktlint/assets/5195292/f2ddec9e-68ca-443c-835c-426cbd4a1ba3)


The resulting code now looks as follows:
![Screenshot 2023-12-19 at 21 26 41](https://github.com/pinterest/ktlint/assets/5195292/571065b0-a6ff-4989-87cf-ecf0948b6e73)

Please note that the comments are no longer related to the original parameter.


> It would be nice to see the mention of the new value-argument-comment and value-parameter-comment rules
> 
> in docs: https://pinterest.github.io/ktlint/1.1.0/rules/standard/
> in release notes: https://github.com/pinterest/ktlint/releases/tag/1.1.0, ideally including some kind of migration guide?

Yes indeed, this should have been documented more clearly. I will fix it.

**Comment by Spikhalskiy** [comment]
@paul-dingemans Thank you for the context. Just a thought. It may be beneficial to make a new section in the documentation that would share knowledge about why a rule is dependent on others. Because it looks like some connections are not obvious and I'm sure more users will be asking you or themselves the same questions.

**Comment by scottkennedy** [comment]
I vehemently agree with this report.  At first, I thought I'd just move all the comments in our codebase, but there are hundreds, and they are much more readable on the same line.  I really want to disable `value-argument/parameter-comment`, but I want `argument-list-wrapping` and `function-signature`.

I do not think these things should be coupled.  I don't think anyone on my team has ever moved a parameter using that UI, and I'd rather eat the cost of manually moving a comment in that case.

**Comment by paul-dingemans** [comment]
> I vehemently agree with this report. At first, I thought I'd just move all the comments in our codebase, but there are hundreds, and they are much more readable on the same line. I really want to disable `value-argument/parameter-comment`, but I want `argument-list-wrapping` and `function-signature`.
> 

Nice things sometimes come at cost. But I do hear your argument that refactoring hundreds of lines of code is not an option for you. 

> I do not think these things should be coupled. 

It is technical complex and most likely it will results in lots of other problems when trying to fix this. You are free to submit a PR which handles comments in value argument lists correctly.

> I don't think anyone on my team has ever moved a parameter using that UI, and I'd rather eat the cost of manually moving a comment in that case.

This is not a relevant argument. Even if ktlint could handle comments correctly, the problem inside IntelliJ IDEA still exists. The IntelliJ IDEA example was simple to  illustrate that even in Jetbrains products this kind of code can lead to unpredictable results.

Next best thing that I can think of, is that the rules that currently have a dependency on any of the comment rules just simply ignore elements containing unexpected comments. But this will also lead to unpredictable results as devs will not understand why
```
        val object = Object(
            someParam1 = 10000, someParam2 = 100, // comment here about why
            sameParam3 = emptyList(),
        )
```
is not being wrapped, while it does work for:
```
        val object = Object(
            someParam1 = 10000, someParam2 = 100,
            sameParam3 = emptyList(),
        )
```

**Comment by scottkennedy** [comment]
> > I vehemently agree with this report. At first, I thought I'd just move all the comments in our codebase, but there are hundreds, and they are much more readable on the same line. I really want to disable `value-argument/parameter-comment`, but I want `argument-list-wrapping` and `function-signature`.
> 
> Nice things sometimes come at cost. But I do hear your argument that refactoring hundreds of lines of code is not an option for you.

That's not the argument.  The argument is
> they are much more readable on the same line

I've refactored the codebase many times when ktlint rule enforcement results in "better" code.  This one does not.

I don't even understand why the comment comes into play with
```
        val object = Object(
            someParam1 = 10000, someParam2 = 100, // comment here about why
            sameParam3 = emptyList(),
        )
```

You have multiple lines, and two params on one line.  That's not allowed.  Leave it to the developer to fix it.

**Comment by Spikhalskiy** [comment]
I think the main problem with this chain disabling from the user perspective is the following:

1. We want to have comments for the arguments after the commas. Which is, indeed, strictly speaking, grammatically incorrect. Such comments do belong to the next argument from the parser's perspective.
2. We disable Rule 1 hoping to just leave such comments be. 
3. But that leads to disabling a higher-level Rule 2.
4. Ok, it's understandable that Rule 2 can't work correctly without applying Rule 1.
5. But what about all the places that are compliant with Rule 1 and we still want Rule 2 to be applied to them? And it can be applied because they are compliant with Rule 2 prerequisites.

I don't want to disable argument-list-wrapping for the whole project for all function/method invocations just because some minority of these calls are not compliant with value-argument-comment.

**Comment by paul-dingemans** [comment]
> I think the main problem with this chain disabling from the user perspective is the following thinking process:
> 
> 1. We want to have comments for the arguments after the commas. Which is, indeed, strictly speaking, grammatically incorrect. Such comments do belong to the next argument from the parser's perspective.
> 2. We disable Rule 1 for the whole project hoping to just leave such comments be.
> 3. But that forces disabling a higher-level Rule 2 for the whole project.
> 4. Ok, it's understandable that Rule 2 can't work correctly with invocations with such comments without applying Rule 1.
> 5. But what about all the places that are already compliant with Rule 1 and we still want Rule 2 to be applied to them?  Rule 2 doesn't have to be disabled project-wise, it can just skip occurrences non-compliant with its dependencies. We still generally want Rule 2.
> 
> I don't want to disable argument-list-wrapping for the whole project for all function/method invocations just because some minority of these calls are not compliant with value-argument-comment.

I think this matches with what I suggested:

> Next best thing that I can think of, is that the rules that currently have a dependency on any of the comment rules just simply ignore elements containing unexpected comments.

**Comment by bay73** [comment]
@paul-dingemans 
Explanation why these rules are introduced seems reasonable, but it really sad that technical limitation and incorrect behavior of some tools leads to decreasing of code readability. I think linter is suppose to make code more comprehensible and not vice versa.

**Comment by paul-dingemans** [comment]
> Explanation why these rules are introduced seems reasonable, but it really sad that technical limitation and incorrect behavior of some tools leads to decreasing of code readability. I think linter is suppose to make code more comprehensible and not vice versa.

I fail to see the point you are trying to make.

**Comment by Spikhalskiy** [comment]
@bay73 

> incorrect behavior of some tools

If you reference the IntelliJ behavior of treating a trailing comment after a comma as part of the next parameter, it's not an incorrect behavior. You may find it inconvenient or not to your taste. But technically from a language grammar perspective, a comma is a separator of parameters and arguments. Such trailing comments do belong to parameters on the next line according to the language spec. 
Tools should behave by respecting correctness and specification first, convenience, taste, and readability second.

**Comment by bay73** [comment]
@Spikhalskiy 
I see, that you failed to understand. You referring again to "technically from a language perspective" and this is exactly what is wrong - technical limitations should not be an argument when we are taking care about readability. Readability is for human and if technical limitations hinder human needs, then technical limitations should be changed, not the human needs.
The ultimate goal of linter is to make it easier for human to work with the code, not for computer or compiler. And now instead of moving to this goal linter becoming a tool with which I have to fight (of suppress) to have readable code.

**Comment by Spikhalskiy** [comment]
@bay73 I do understand your point of view, but I show that it may not be the only one. 

For example, from my point of view, the purpose of a linter is to make code easier to read and "prettier", but inside the box of the language specification and grammar. I wouldn't want a linter to start bending boundaries defined by them.

If the linter decides that it knows better, and steps out of the rules of the language grammar, it will unavoidably run into incompatibilities with other tools that do it differently. As Paul pointed out in the example with Intelij refactoring.

**Comment by YSakhno** [comment]
Frankly speaking your reasoning about "breaking the IDE / unexpected behavior in IntelliJ IDEA" does not work, because the refactoring does not move the comments even in the case when comments are written _before_ each argument.

Before the refactoring:

```kotlin
data class Foo(val someParam1: Int, val someParam2: Int, val someParam3: List<String>)

val foo = Foo(
    // some comment here about why 1
    someParam1 = 10_000,
    // some comment here about why 2
    someParam2 = 100,
    // some comment here about why 3
    someParam3 = emptyList(),
)
```

After the "Change signature" refactoring:

```kotlin
data class Foo(val someParam1: Int, val someParam3: List<String>, val someParam2: Int)

val foo = Foo(
    // some comment here about why 1
    someParam1 = 10_000,
    // some comment here about why 2
    someParam3 = emptyList(),
    // some comment here about why 3
    someParam2 = 100,
)
```

This is not to mention that the rule is also reported for the cases when arguments are passed in place of `vararg` parameters (i.e. to functions like `listOf()`, `setOf()` and the like), where **no** signature refactorings are ever possible anyway (and so nothing can potentially ever "break").

The syntax and formatting in the snippets above are "allowed" by the latest ktlint version (**1.1.1**), BTW. You've rolled out a half-baked raw rule, which clearly had to be hidden in the experimental section (if even implemented at all), and now after updating the tool version, I've got bunch of rule violations for the code that had "perfectly good" style according to the older version of ktlint.

P.S. The documentation for the rules is also half-baked, because it seems to repeat word-for-word the explanation for rules `value-argument-comment` and `value-parameter-comment`.

**Comment by paul-dingemans** [comment]
> Frankly speaking your reasoning about "breaking the IDE / unexpected behavior in IntelliJ IDEA" does not work, because the refactoring does not move the comments even in the case when comments are written _before_ each argument.

Nice find! I never noticed that.

I will ignore your other remarks as I take offense about using the phrase "half-baked". If you can do better, please raise a PR to fix.

**Comment by YSakhno** [comment]
Sorry if I offended you, but whether you find the language offensive or not, the quality of the current implementation remains to be what it is. I should stress here that I do not mean to berate the quality of the _entire tool_, which although occasionally has some problematic spots from version to version, largely is very mature, robust and tremendously useful.

I can only raise a PR that reverts this particular rule (`value-argument-comment`), as I find it useless (for me, so I have to turn it off anyway). It seems some other people have issues with it too. Also it contributes to "bike-shedding" (just look at the discussion above) — the practice this very tool claims to be against (or at least aims to alleviate to some extent).

I am speculating here, but I would venture a guess, that the majority of people who have no issue with this rule are those, who do not write comments in the code (at all, of any kind) to begin with.

**Comment by paul-dingemans** [comment]
> I can only raise a PR that reverts this particular rule (value-argument-comment), as I find it useless (for me, so I have to turn it off anyway). It seems some other people have issues with it too. Also it contributes to "bike-shedding" (just look at the discussion above) — the practice this very tool claims to be against (or at least aims to alleviate to some extent).

Thanks for clarifying. In your original comment it was not clear to me that you were referring to the `value-argument-comment`. I had another look at how the comments are actually parsed, and found that I made a mistake there. Some comments in a `value-argument-list` or `value-parameter-list` are siblings of the `value-argument` or `value-parameter`. Such comments should not be forbidden by the `value-argument-comment` and `value-parameter-comment` rules. See #2519.

**Comment by brill-touchbistro** [comment]
This seems like it already long past, but to add my 0.02:
I got here for the same reason the original issue was raised. 

IMO The code is much more readable with an inline comment, and that to me is far more important then any technically pedantic reasoning for not allowing it to be independently configured. 

Allow the option to be independently configurable, whether it is enabled by default or not, so that choices can be made.
There is no reason not to do that.

**Comment by paul-dingemans** [comment]
> This seems like it already long past, but to add my 0.02:

Maybe you can put some more effort in your remark. Did you check against latest release? What exactly is not correct? What is your configuration? Please create a new issue, if applicable.
